app/javascript/components/patient/vaccines/VaccineModal.js

```diff
@@ -0,0 +1,244 @@
+        ? this.props.currentVaccineData.group_name
+        : this.props.group_name_options
+        ? this.props.group_name_options[0]
+        : null,
+      administration_date: this.props.currentVaccineData?.administration_date,
+      dose_number: this.props.currentVaccineData?.dose_number,
+      notes: this.props.currentVaccineData?.notes,
+      // Sorting so that the blank option shows up at the top
+      sorted_dose_number_options: this.props.dose_number_options ? this.props.dose_number_options.sort() : [],
+      isValid: false,
+    };
+  }
+
+  componentDidMount() {
+    this.checkIfValid();
+  }
+
+  /**
+   * Data is valid and can be saved if there is both a group name and product name selected.
+   */
+  checkIfValid = () => {
+    const isValid = this.state.group_name && this.state.product_name;
+    this.setState({ isValid });
+  };
+
+  handleProductNameChange = data => {
+    this.setState({ product_name: data.value }, () => {
+      this.checkIfValid();
+    });
+  };
+
+  handleGroupNameChange = data => {
+    // Resets product name if selected because product name options are dependent on group name
+    this.setState({ group_name: data.value, product_name: null }, () => {
+      this.checkIfValid();
+    });
+  };
+
+  handleAdministrationDateChange = newDate => {
+    this.setState({ administration_date: newDate }, () => {
+      this.checkIfValid();
+    });
+  };
+
+  handleDoseNumberChange = data => {
+    this.setState({ dose_number: data.value }, () => {
+      this.checkIfValid();
+    });
+  };
+
+  handleNotesChange = event => {
+    const val = event.target.value;
+    this.setState({ notes: val }, () => {
+      this.checkIfValid();
+    });
+  };
+
+  /**
+   * Gets the list of product name options based on the currently selected vaccine group.
+   * Appends the additional product name options that are available no matter the group.
+   */
+  getProductNameOptions = () => {
+    const product_names = this.props.vaccine_mapping ? this.props.vaccine_mapping[this.state.group_name]?.vaccines?.map(vaccine => vaccine.product_name) : [];
+    return product_names ? product_names.concat(this.props.additional_product_name_options) : [];
+  };
+
+  render() {
+    const defaultGroupNameOption = this.props.group_name_options ? this.props.group_name_options[0] : '';
+    return (
+      <Modal size="lg" show centered onHide={this.props.onClose}>
+        <h1 className="sr-only">{this.props.title}</h1>
+        <Modal.Header>
+          <Modal.Title>{this.props.title}</Modal.Title>
+        </Modal.Header>
+        <Modal.Body>
+          <Form>
+            <Row>
+              <Form.Group as={Col}>
+                <Form.Label htmlFor="group-name-select" className="nav-input-label">
+                  Vaccine Group
+                </Form.Label>
+                <Select
+                  inputId="group-name-select"
+                  name="group_name"
+                  defaultValue={
```

Comment:
Can we break this label/value logic out into its own `getFormattedOptions` function and then find the proper values from that?  See how it's done in the advanced filter, I think it's a lot cleaner and easier to understand

Reply:
Yeah great idea - done here: https://github.com/SaraAlert/SaraAlert/pull/721/commits/1af7e86f6a26f91f7b2b6563d00d493acf16da44#diff-182e13a82e02db14eb5222d1eb289b14639dc17acd927608ff7e8b032cfb9af8R68
I did this for the `options` attribute, not `defaultValue` since that wasn't always the same pattern. I believe this should resolve all your other comments as well, if not let me know.